### PR TITLE
Use a metaclass on GenericMap to automatically register subclasses.

### PR DESCRIPTION
**Pull request by Cadair**
This uses the concept of metaclasses to automatically register the subclasses of `GenericMap` with `Map` upon creation (import) of the classes. 
This removes the need to explicitly register your `GenericMap` subclass with `Map` (although this is still technically possible if you are **totally** insane).

For more information on metaclasses see this rather interesting stackoverflow page: https://stackoverflow.com/questions/100003/what-is-a-metaclass-in-python

One implementation detail here is that the registry is maintained inside the namespace of `mapbase` this is done because if the `GenericMapMeta` class tried to import the registry stored in the instance `Map` in `map_factory` a horrible import loop was formed, where `mapbase` needed to import `map_factory` and vice-a-versa.

@rhewett seeing how this is your code I am messing with here, what do you think?

This is missing any explicit tests (although the existing tests for `Map` are testing it) and a changelog, but I would like some feedback.

### CODE REVIEW

**Comment by rhewett** [comment]
:+1:

The only comment I have is that it forces the "check" function to have a specific name.  This is not necessarily a bad thing, it is just a change.  We were not using that "feature" anyways, so it should not be a problem.

**Comment by dpshelio** [comment]
Wow! are we in the 1% mentioned in the SO post? I'm happy with it... but it goes way far of my comprehension.  :+1:  but I would like that someone else that understands it approves it too (I don't like low :bus: factors)

**Comment by ayshih** [comment]
Check what happens to the NDData metaclass after this is done

**Comment by Cadair** [comment]
@ayshih so just subclassing our meta from NDDataBase doesn't work.

**Comment by Cadair** [comment]
@ehsteve did you have comments on this?

**Comment by ayshih** [comment]
:-1:

=P

**Comment by Cadair** [comment]
@ayshih updated.

File: doc/source/code_ref/map.rst [hunk 1]
```diff
@@ -87,16 +90,21 @@ Instrument Map Classes
 Writing a new Instrument Map Class
 ----------------------------------
 
-Map classes can be registered with the Map factory, even if the new class is not
-officially part of SunPy.  This is good for prototyping new instruments.  For
-example, to add a Map type for a future instrument, consider this code skeleton::
+Any subclass of `~sunpy.map.GenericMap` which defines a method named
+``is_datasource_for`` will automatically be registered with the ``Map`` factory.
+This makes it straightforward to define your own `~sunpy.map.GenericMap`
+subclass for a new instrument or a custom data source.
```
Comment (ehsteve) [inline doc/source/code_ref/map.rst]: mention simulations

File: doc/source/code_ref/map.rst [hunk 2]
```diff
@@ -87,16 +90,21 @@ Instrument Map Classes
...
+These classes do not have to be part of SunPy for this to work, as
```
Comment (ehsteve) [inline doc/source/code_ref/map.rst]: this is confusing.

File: doc/source/code_ref/map.rst [hunk 3]
```diff
@@ -87,16 +90,21 @@ Instrument Map Classes
 Writing a new Instrument Map Class
 ----------------------------------
 
-Map classes can be registered with the Map factory, even if the new class is not
-officially part of SunPy.  This is good for prototyping new instruments.  For
-example, to add a Map type for a future instrument, consider this code skeleton::
+Any subclass of `~sunpy.map.GenericMap` which defines a method named
+``is_datasource_for`` will automatically be registered with the ``Map`` factory.
```
Comment (ehsteve) [inline doc/source/code_ref/map.rst]: explain how this dynamically add capability to your SunPy.

File: doc/source/code_ref/map.rst (hunk 2 above)
Comment (ehsteve) [inline doc/source/code_ref/map.rst]: suggest to remove.

File: doc/source/code_ref/map.rst (hunk 3 above)
Comment (ehsteve) [inline doc/source/code_ref/map.rst]: define what ``is_datasource_for`` and add link to it.

File: doc/source/code_ref/map.rst [hunk 4]
```diff
@@ -107,11 +115,19 @@ example, to add a Map type for a future instrument, consider this code skeleton:
...
             return header.get('instrume', '').startswith('FUTURESCOPE')
 
-Then, to be able to instantiate a FutureMap using the Map() factory, one must
-register the FutureMap type with the factory::
 
-    sunpy.map.Map.register(FutureMap, FutureMap.is_datasource_for)
+This class will now be available through the ``Map`` factory as long as this
+class has been defined, i.e. imported into the current session.
+
+If you do not want to create a method named ``is_datasource_for`` (or for some
```
Comment (Cadair) [inline doc/source/code_ref/map.rst]: explicit name.

**Comment by ehsteve** [comment]
:+1: if documentation gets fixed or if not then an issue is added to fix it later.

**Comment by Cadair** [comment]
all faliures are expected or ok.

**Pull request merged**